lightning/src/routing/router.rs

```diff
@@ -1785,10 +1780,57 @@ fn add_random_cltv_offset(route: &mut Route, payment_params: &PaymentParameters,
 	}
 }
 
+/// Build a route from us (payer) with the given hops ending at the target node (payee).
+///
+/// Re-uses logic from `find_route`, so the restrictions described there also apply here.
+pub fn build_route_from_hops<L: Deref>(
+	our_node_pubkey: &PublicKey, hops: &[PublicKey], route_params: &RouteParameters, network: &NetworkGraph,
+	logger: L, random_seed_bytes: &[u8; 32]) -> Result<Route, LightningError>
+where L::Target: Logger {
+	let network_graph = network.read_only();
+	let mut route = build_route_from_hops_internal(
+		our_node_pubkey, hops, &route_params.payment_params, &network_graph,
+		route_params.final_value_msat, route_params.final_cltv_expiry_delta, logger, random_seed_bytes)?;
+	add_random_cltv_offset(&mut route, &route_params.payment_params, &network_graph, random_seed_bytes);
+	Ok(route)
+}
+
+fn build_route_from_hops_internal<L: Deref>(
+	our_node_pubkey: &PublicKey, hops: &[PublicKey], payment_params: &PaymentParameters,
+	network_graph: &ReadOnlyNetworkGraph, final_value_msat: u64, final_cltv_expiry_delta: u32,
+	logger: L, random_seed_bytes: &[u8; 32]) -> Result<Route, LightningError> where L::Target: Logger {
+
+	struct HopScorer<'a> { hops: &'a [PublicKey] }
+
+	impl Score for HopScorer<'_> {
+		fn channel_penalty_msat(&self, _short_channel_id: u64, source: &NodeId, target: &NodeId,
+			_usage: ChannelUsage) -> u64
+		{
+			for i in 0..self.hops.len()-1 {
+				let cur_id = NodeId::from_pubkey(&self.hops[i]);
```

Comment:
Another idea would be to hold the first hop as a separate field (which would avoid the `Vec` allocation in `concat`) as well as keeping track of `cur_id` outside the loop (moving to `next_id` at the end of each iteration). That way, you only need to `n` calls to `NodeId::from_pubkey` instead of `2n - 1`

Reply:
Great idea, did that in https://github.com/lightningdevkit/rust-lightning/pull/1491/commits/ba48ac88435bb3e6cda89d393944bb0f62d9c066.